[MRG] Parameter for stacking missing indicator into imputer

DanilBaibak opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Fixes #11886 
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


A new parameter `add_indicator` was added to [SimpleImputer](https://scikit-learn.org/dev/modules/generated/sklearn.impute.SimpleImputer.html) that allow simply stacking a [MissingIndicator](https://scikit-learn.org/dev/modules/generated/sklearn.impute.MissingIndicator.html) transform into the output of the imputer's transform.




<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

DanilBaibak committed:
Added the parameter to SimpleImputer that allow to stack MissingIndicator transform into it

DanilBaibak committed:
Added tests for the new functionality

DanilBaibak committed:
Updated docs

DanilBaibak committed:
Fixed code style issues

amueller commented on sklearn/impute.py:
```diff
@@ -406,6 +421,10 @@ def transform(self, X):
 
+        # stacks a MissingIndicator transform into the output of the imputer
+        if self.add_indicator:
+            X = np.concatenate((X, X_trans), axis=1)
```
What if X is sparse?

DanilBaibak replied:
Good point. Thanks.

DanilBaibak committed:
Use sparse.hstack for stacking the sparse matrices

DanilBaibak committed:
Added tests for the case of the sparse matrices

DanilBaibak committed:
Fixed tests

amueller commented:
Sweet. Can you add an entry to the 0.21 whatsnew? Thanks!

DanilBaibak committed:
Added an entry to the 0.21 what's new

DanilBaibak committed:
Resolve conflicts after merge with master

jeremiedbb commented on sklearn/impute.py:
```diff
@@ -133,6 +133,10 @@ class SimpleImputer(BaseEstimator, TransformerMixin):
         - If X is not an array of floating values;
         - If X is encoded as a CSR matrix.
 
+    add_indicator : boolean, optional (default=False)
```
When `copy=False` and `add_indicator=True`, a copy is still made. We should document it in the copy description.

jeremiedbb commented on sklearn/tests/test_impute.py:
```diff
@@ -635,3 +635,57 @@ def test_inconsistent_dtype_X_missing_values(imputer_constructor,
+    X_trans = imputer.fit_transform(X)
+
+    assert_array_equal(X_trans, X_true)
+    assert_array_equal(imputer.statistics_, np.array([3., 2., 5.]))
```
You should use `assert_array_almost_equal` when comparing floating values.

jeremiedbb reviewed (commented):
This is a nice feature ! Thanks. A few comments below.

DanilBaibak committed:
Improved tests

DanilBaibak commented on doc/whats_new/v0.21.rst:
```diff
@@ -138,6 +138,13 @@ Support for Python 3.4 and below has been officially dropped.
+- |Feature| The :class:`impute.SimpleImputer` has a new parameter
+  ``'add_indicator'``, which simply stacks a :class:`impute.MissingIndicator`
+  transform into the output of the imputer's transform. :issue:`11886` by
+  :user:`Danylo Baibak <DanilBaibak>`.
```
Thanks, I was a bit confused with `issue`/`PR` :)

DanilBaibak committed:
Code review improvements

DanilBaibak committed:
Fixed what's new doc

jnothman commented on sklearn/impute.py:
```diff
@@ -161,12 +165,16 @@ class SimpleImputer(BaseEstimator, TransformerMixin):
+        self.add_indicator = add_indicator
+
+        if self.add_indicator:
+            self.indicator_ = MissingIndicator(missing_values=missing_values)
```
We don't do anything like this in constructor usually. set_params can change the value of add_indicator between construction and fit

jeremiedbb commented on sklearn/tests/test_impute.py:
```diff
@@ -657,7 +657,7 @@ def test_imputation_add_indicator(marker):
 
     assert_array_equal(X_trans, X_true)
-    assert_array_equal(imputer.statistics_, np.array([3., 2., 5.]))
+    assert_allclose(imputer.statistics_, np.array([3., 2., 5.]))
```
`assert_array_equal(X_trans, X_true)` needs `assert_all_close` too.

jeremiedbb commented on sklearn/tests/test_impute.py:
```diff
@@ -686,6 +685,6 @@ def test_imputation_add_indicator_sparse_matrix(arr_type):
     X_trans = imputer.fit_transform(X_sparse)
 
-    assert_true(sparse.issparse(X_trans))
+    assert sparse.issparse(X_trans) is True
```
you don't need to add `is True`.

DanilBaibak committed:
Moved MissingIndicator initialization into the fit method

DanilBaibak committed:
Improved tests

jeremiedbb commented:
Hum I found a case when it will fail. When one column is full of missing value, the SimpleImputer will drop that column (except for "constant" strategy) and the MissingIndicator will have a feature mismatch at transform time.

First, you can add it to the test, i.e. add one column full of missing values, to make the test cover more cases. 

Then, to fix this, one possibility is to not fit the MissingIndicator in fit but fit_transform it instead in transform. The MissingIndicator does not hold any information that the SimpleImputer already have, so it does not even need to be an attribute of the SimpleImputer. @amuller @jnothman  what do you think ?

jeremiedbb commented on sklearn/impute.py:
```diff
@@ -258,6 +264,10 @@ def fit(self, X, y=None):
                                                fill_value)
 
+        if self.add_indicator:
+            self.indicator_ = MissingIndicator(missing_values=self.missing_values)
```
pep8, this line is too long. I wonder why the CI did not fail.

DanilBaibak committed:
Added fix for the case when one column is totally missing

DanilBaibak committed:
Added more tests for the case with missing column

DanilBaibak commented:
@jeremiedbb, good catch with fully missing column!  I adjusted code and added tests. What do you think about it now?

amueller commented:
I think ``transform`` shouldn't modify the estimator. Why can't we fit in fit and call transform after dropping columns?

I don't like the current solution since it adds a constant 1 column which is not useful imho.

DanilBaibak commented:
@amueller and @jeremiedbb, I see your point and agree. But the current solution does the same if you do it like this:
```
pipeline = make_pipeline(
    make_union(
        SimpleImputer(missing_values=marker),
        MissingIndicator(missing_values=marker)
    )
)

pipeline.fit_transform(X_test)
``` 
In theory, if we drop the column full of 1, people who now use `make_union` can affected. What do you think?

DanilBaibak committed:
Improved the tests

amueller commented:
That's a good point. We need to decide to either:

1) make it inconsistent and drop the 1s here
2) keep the redundant 1s in both places
3) change the behavior in MissingIndicator to always drop constant columns.

I'm somewhat leaning towards 3, though it requires a deprecation cycle.

DanilBaibak commented:
Seems, auto drop for a constant column not always good choice. Here's a small sample:
```
X_train = np.array([
    [1, 1, np.nan],
    [np.nan, 2, 6],
    [2, np.nan, 3],
    [3, 3, 9]
])

X_test = np.array([
    [np.nan, 1, 5],
    [np.nan, 2, np.nan],
    [np.nan, np.nan, 3]
])

pipeline = make_pipeline(
    make_union(
        SimpleImputer(),
        MissingIndicator()
    )
)
pipeline.fit(X_train)

pipeline.transform(X_train)
pipeline.transform(X_test)
```
For `X_test` we have a whole column of 1, but not for `X_train`.

I would vote for the # 2, because, to be honest, it's hard to imagine, that users will try to work with a dataset, that contains whole column of empty values. It would be unuseful even with "constant" strategy for SimpleImputer.

jnothman commented:
Yes, as a rule you can't change anything on the basis of the test data.

jeremiedbb commented:
> For X_test we have a whole column of 1, but not for X_train.

It's not a problem. In that case, we don't drop the column from X_test. It's the same in the SimpleImputer. If X_train has a column with not all missing values, you can fit it and get the mean and if X_test has this column full of missing values, we impute it with the mean previously computed.

It's the fit that needs to be changed if we decide to do (3.), not the transform.

DanilBaibak committed:
Adjusted MissingIndicator to always drop constant columns

jeremiedbb commented on sklearn/impute.py:
```diff
@@ -531,12 +554,42 @@ def _get_missing_features_info(self, X):
+
+        if is_sparce:
+            imputer_mask = imputer_mask.toarray()
+
```
This conversion is not a good idea and might lead to memory error on large matrices.

A better way would be to convert it in `csc` format and then keep the non full missing columns. Here is the way to do it:
```python
if self.sparse:
    sparse_format = imputer_mask.format
    imputer_mask = csc_matrix(imputer_mask)

non_fully_missing_indexes = np.where(
    features_with_missing != imputer_mask.shape[0])[0]
imputer_mask = imputer_mask[:, non_fully_missing_indexes]
features_with_missing = features_with_missing[non_fully_missing_indexes]

if self.sparse:
    imputer_mask = csr_matrix(imputer_mask) if sparse_format is 'csr' else imputer_mask
```

jnothman commented on doc/whats_new/v0.21.rst:
```diff
@@ -138,6 +138,13 @@ Support for Python 3.4 and below has been officially dropped.
+.....................
+- |Feature| The :class:`impute.SimpleImputer` has a new parameter
+  ``'add_indicator'``, which simply stacks a :class:`impute.MissingIndicator`
+  transform into the output of the imputer's transform. :issue:`12583` by
```
perhaps add "allowing a predictive estimator to account for missingness"

jnothman commented on sklearn/impute.py:
```diff
@@ -131,7 +131,12 @@ class SimpleImputer(BaseEstimator, TransformerMixin):
+        - If add_indicator=True.
+
+    add_indicator : boolean, optional (default=False)
+        If True, a MissingIndicator transform will stack into output
```
put MissingIndicator in backticks, please

jnothman commented on sklearn/impute.py:
```diff
@@ -131,7 +131,12 @@ class SimpleImputer(BaseEstimator, TransformerMixin):
+
+    add_indicator : boolean, optional (default=False)
+        If True, a MissingIndicator transform will stack into output
+        of the imputer's transform.
```
add "allowing a predictive estimator to account for missingness"

jnothman commented on sklearn/impute.py:
```diff
@@ -531,12 +554,42 @@ def _get_missing_features_info(self, X):
+        """
+        In case we have a column with totally missing  values, drop it,
+        because constant column is not informative.
+        :param imputer_mask:
```
this should be numpydoc

jnothman commented on sklearn/impute.py:
```diff
@@ -519,6 +539,9 @@ def _get_missing_features_info(self, X):
                 np.flatnonzero(np.diff(missing_values_mask.indptr))
                 if missing_values_mask.format == 'csc'
                 else np.unique(missing_values_mask.indices))
+            imputer_mask, features_with_missing = self._drop_constant_columns(
```
This changes public API doesn't it? Doesn't this correspond to adding a different value for the `MissingIndicator.features` parameter?

jeremiedbb replied:
Yes, now columns with only missing values are dropped, as in the `SimpleImputer`. The current options for the `features` parameter are keep all columns or drop columns with no missing.

We could make it another option but we don't in the imputer and columns full of missing bring no information.

jnothman replied:
Well, I'm not convinced we should be changing it without warning. It won't affect usage very often, though.

amueller replied:
yeah looks like it would require a deprecation cycle. If we add an option and pass the option in ``SimpleImputer`` then it's unlikely it'll affect a lot of users, I think.
Most people will probably use the convenience in ``SimpleImputer``.

jnothman replied:
It's unlikely to affect a lot of users in any case... It's dealing with extremely degenerate data

jnothman reviewed (commented):
Sorry for coming late to the party. Not yet looked at tests. Thanks!

DanilBaibak committed:
Fixed issue with indexes for the fully missing columns

DanilBaibak committed:
Improved the tests

DanilBaibak committed:
Improved the descriptions

DanilBaibak committed:
Fixed tests

jnothman commented on doc/modules/impute.rst:
```diff
@@ -30,7 +30,7 @@ that contain the missing values::
     >>> imp = SimpleImputer(missing_values=np.nan, strategy='mean')
     >>> imp.fit([[1, 2], [np.nan, 3], [7, 6]])       # doctest: +NORMALIZE_WHITESPACE
-    SimpleImputer(copy=True, fill_value=None, missing_values=nan, strategy='mean', verbose=0)
+    SimpleImputer(add_indicator=False, copy=True, fill_value=None, missing_values=nan, strategy='mean', verbose=0)
```
seeing as you already have `NORMALIZE_WHITESPACE`, please wrap this to two lines to avoid exceeding 80 characters per line.

jnothman commented on sklearn/impute.py:
```diff
@@ -131,7 +131,13 @@ class SimpleImputer(BaseEstimator, TransformerMixin):
+
+    add_indicator : boolean, optional (default=False)
+        If True, a `MissingIndicator` transform will stack into output
+        of the imputer's transform. That allowing a predictive estimator
```
That allowing -> This allows

jnothman commented on sklearn/impute.py:
```diff
@@ -131,7 +131,13 @@ class SimpleImputer(BaseEstimator, TransformerMixin):
+    add_indicator : boolean, optional (default=False)
+        If True, a `MissingIndicator` transform will stack into output
+        of the imputer's transform. That allowing a predictive estimator
+        to account for missingness.
```
add "despite imputation"

jnothman commented on sklearn/tests/test_impute.py:
```diff
@@ -635,3 +635,55 @@ def test_inconsistent_dtype_X_missing_values(imputer_constructor,
+    X_trans = imputer.fit_transform(X)
+
+    assert_allclose(X_trans, X_true)
+    assert_allclose(imputer.statistics_, np.array([3., 2., 5., np.nan, 2.5]))
```
The invariant you want to hold is that the attributes of SimpleImputer(add_indicator=True) are identical to those of SimpleImputer(add_indicator=False), and their transform is identical except for some added columns. Ideally the test would be written in a way that expresses that invariant rather than repeating the expected values as it does here.

jnothman commented on sklearn/impute.py:
```diff
@@ -551,13 +604,15 @@ def fit(self, X, y=None):
         self : object
             Returns self.
         """
+        self.fully_missing_indexes_ = None
```
indexes -> indices

jnothman commented on sklearn/impute.py:
```diff
@@ -537,6 +558,38 @@ def _get_missing_features_info(self, X):
+            sparse_format = X.format
+            X = X.toarray()
+
+        if not self.fully_missing_indexes_:
```
If fullly_missing_indexes_ is an array, this will break. Please test.

jnothman commented on sklearn/impute.py:
```diff
@@ -600,6 +655,7 @@ def transform(self, X):
         X = check_array(X, accept_sparse=('csc', 'csr'),
                         force_all_finite=force_all_finite)
         _check_inputs_dtype(X, self.missing_values)
+        X = self._drop_constant_columns(X)
```
This code is deceptive. It looks like it will use statistics from X to drop columns. Rather, perhaps, pass fully_missing_indices_ here.

jnothman commented on sklearn/impute.py:
```diff
@@ -537,6 +558,38 @@ def _get_missing_features_info(self, X):
 
         return imputer_mask, features_with_missing
 
+    def _drop_constant_columns(self, X):
```
"constant" is poorly descriptive. You mean "all-missing"

jnothman commented on sklearn/impute.py:
```diff
@@ -537,6 +558,38 @@ def _get_missing_features_info(self, X):
+            X = X.toarray()
+
+        if not self.fully_missing_indexes_:
+            ix = _get_mask(X, self.missing_values)
```
Is there a way we can simply reuse an already-calculated mask? This seems to be unnecessary duplication of work. Can we just look for statistics_ with value NaN?

jeremiedbb replied:
We can't just look for `statistics_` since for the `MissingIndicator` we want to drop the columns with no missing values (which is it's default behavior). This info can't be extracted from `statistics_`.

We could swith to the `features='all'` behavior but I think it's better to only keep the relevant features.

To avoid recomputing the mask, we could make `MissingIndicator` to accept a precomputed mask. What do you think about that ?

jnothman commented on sklearn/impute.py:
```diff
@@ -551,13 +604,15 @@ def fit(self, X, y=None):
         X = check_array(X, accept_sparse=('csc', 'csr'),
                         force_all_finite=force_all_finite)
         _check_inputs_dtype(X, self.missing_values)
+        X = self._drop_constant_columns(X)
```
Doing this by default breaks backwards compatibility. This is a serious issue with the PR as it stands.

I agree that this is sensible behaviour, but it is also handling a rare edge case and I would rather waste little code on it. My main concern for a column with all NaNs is that we do not raise an obscure error message. I don't really care how it's handled as long as it's backwards compatible.

DanilBaibak committed:
Code review improvements regarding naming and descriptions

jnothman commented:
Please merge in the latest master, also. "ci/circleci: python2" should no longer be run.

glemaitre commented:
We also need to solve the issue #13035 and ensure that there is a test working with non-numeric data in this PR.

jeremiedbb committed:
option to drop full missing columns (dense)

jeremiedbb committed:
deal with sparse

jeremiedbb committed:
what' new

jeremiedbb commented:
#13035  has been fixed by #13046

I made #13491 to add an option to drop fully missing columns in a backward compatible way. It should help this PR move forward.

@DanilBaibak are you still willing to work on this ?

jeremiedbb committed:
tst

DanilBaibak committed:
Resolved conflicts after merge with master

jeremiedbb committed:
add test for explicit zeros

jeremiedbb committed:
what's new

jnothman commented:
Don't agonise over efficiency in the first instance... But if we needed
some private implementation of MissingIndicator that accepted a precomputed
mask, that would be okay

DanilBaibak committed:
Improved code quality

jnothman commented:
Danylo, would you be able to merge in and use the changes from #13491 in
your branch? Or would you rather wait until that is approved and merged?

DanilBaibak commented:
Before merge #13491, I have problem with test for the my PR:
`test_estimators[MissingIndicator-check_dict_unchanged]` with message `Estimator changes __dict__ during transform`. Any suggestion, how better fix it?

jeremiedbb committed:
simpler code. fix typos. wording.

jeremiedbb commented:
Before you merge it, I'd advise you remove all changes to the `MissingIndicator` you made in this PR. It'll make the merge a lot easier. And I think it will fix the test, since it seems related to this estimator (it's hard to find why the test fails with this uninformative error message).

DanilBaibak committed:
Reverted changes for the MissingIndicator

DanilBaibak committed:
Resolved conflicts after merge with #13491

DanilBaibak committed:
Set the paramter 'features' for MissingIndicator as 'some-missing' by default in case of stacking SimpleImputer and MissingIndicator

DanilBaibak commented:
@jeremiedbb thanks for the your PR!

I also adjusted SimpleImputer to set the parameter `features` for MissingIndicator as `some-missing` by default in case of stacking SimpleImputer and MissingIndicator. Seems, that should be also updated in the documentation.

jeremiedbb reviewed (commented):
Just a few comments. LGTM once we agree on the `error_on_new` question.
(and I think one comment of Joel about a test hasn't been adressed)

jnothman commented:
Just to let you know. We might want similar in IterativeImputer, once we
are happy with this.

DanilBaibak committed:
Improved code quality

DanilBaibak committed:
Improved the documentation

DanilBaibak committed:
Fixed the line length

DanilBaibak committed:
Added ability to stack IterativeImputer and MissingIndicator

DanilBaibak committed:
Fixed the line length

jeremiedbb commented:
@DanilBaibak I think Joel meant in another PR, once this one is done

jeremiedbb committed:
update docstring + typo

jeremiedbb committed:
Merge remote-tracking branch 'upstream/master' into missing-indicator-drop-full-missing

jeremiedbb committed:
update error_on_new

jeremiedbb committed:
add test

DanilBaibak commented:
Ok! Just added it in hot pursuit :simple_smile:

DanilBaibak committed:
Rollback the changes for the IterativeImputer

DanilBaibak committed:
Resolved conflicts after merge

jnothman commented on sklearn/impute.py:
```diff
@@ -272,6 +282,12 @@ def fit(self, X, y=None):
                                                fill_value)
 
+        if self.add_indicator:
+            self.indicator_ = MissingIndicator(
```
Let's make this private, i.e. `_indicator`. If we make it public it needs to be described in the Attributes section of the docstring, above.

DanilBaibak replied:
Maybe, let's add the description for the `_indicator` in the Attributes section. It's still can be useful to see `features_` of the MissingIndicator (`imputer.indicator_.features_`).

jnothman replied:
> Maybe, let's add the description for the indicator_ in the Attributes section

Okay

jnothman commented on sklearn/impute.py:
```diff
@@ -141,7 +141,16 @@ class SimpleImputer(BaseEstimator, TransformerMixin):
+        - If add_indicator=True.
+
+    add_indicator : boolean, optional (default=False)
+        If True, a `MissingIndicator` transform will stack into output
```
into -> onto

jnothman reviewed (approved):
Looking good following merge of #13491

DanilBaibak commented on sklearn/impute.py:
```diff
@@ -1190,6 +1213,41 @@ def _validate_input(self, X):
+
+        if not self.fully_missing_indices_:
+            ix = _get_mask(X, self.missing_values)
+            self.fully_missing_indices_ = np.where(ix.all(axis=0))[0]
```
True! Seems, you're looking into outdated code

DanilBaibak committed:
Improved the description

DanilBaibak committed:
Added the description for the indicator_ in the Attributes section

DanilBaibak committed:
Updated the description in the Attributes section

DanilBaibak committed:
Resolved the conflicts after merging with master

DanilBaibak committed:
Fixed the tests after merge with master

DanilBaibak committed:
Fixed code style issue

NicolasHug commented on doc/whats_new/v0.21.rst:
```diff
@@ -569,6 +569,14 @@ Support for Python 3.4 and below has been officially dropped.
   affects all ensemble methods using decision trees.
   :issue:`12344` by :user:`Adrin Jalali <adrinjalali>`.
 
+:mod:`sklearn.impute`
```
There's already a `sklearn.impute` section above, please move it there

NicolasHug commented on sklearn/impute.py:
```diff
@@ -272,6 +285,11 @@ def fit(self, X, y=None):
                                                fill_value)
 
+        if self.add_indicator:
+            self.indicator_ = MissingIndicator(
```
`self.indicator_` should be `None` if `self.add_indicator = False`. 

This currently fails:

```
SimpleImputer().fit(X).indicator_
```

with `AttributeError: 'SimpleImputer' object has no attribute 'indicator_'`

Also please update the attribute docstring with "[...] None if add_indicator is False."

NicolasHug commented on sklearn/impute.py:
```diff
@@ -382,6 +399,9 @@ def transform(self, X):
                              % (X.shape[1], self.statistics_.shape[0]))
 
+        if self.add_indicator:
+            X_trans = self.indicator_.transform(X)
```
It feels weird to call this `X_trans` when the actual `X_trans` is named `X`.

Maybe rename this into `X_trans_indicator`

NicolasHug reviewed (commented):
A few comments

DanilBaibak committed:
Code review improvements

jnothman reviewed (approved):
Yes! Thanks.

jnothman commented:
Thanks @banilo!!

DanilBaibak commented:
Glad to help 😊
